### PR TITLE
NoRA

### PR DESCRIPTION
**Pull request by Joluck**
Sharing our recent work: Normalized Low-Rank Adaptation (NoRA). 
Inspired by latent normalization in MLA, we find that normalizing Norm(Ax) improves optimization, but introduces input-dependent nonlinearity and breaks mergeability.
NoRA instead moves normalization to the down-projection itself:
Norm(Ax) → Norm(A)x by normalizing A along the rank dimension.

We further propose Block Identity Matrix Initialization (BIMI) as a simple deterministic realization of NoRA-Init.

https://huggingface.co/papers/2608.31036

### CODE REVIEW

**Comment by Joluck** [comment]
@BenjaminBossan

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for integrating these new options into LoRA.

To see the results, I ran the MetaMathQA/GSM8K benchmark in PEFT. I'm not exactly sure if the experiment is directly comparable to the one in the paper, do you have the code for that somewhere? The end result for me was that the test accuracy for LoRA, NoRA, and BIMI are very close, within the noise range (all rank 32):

- LoRA: [0.478](https://github.com/huggingface/peft/blob/05456023a231632c1de47b23edafa7e5adf86148/method_comparison/MetaMathQA/results/lora--llama-3.2-3B-rank32.json#L334) (with [this config](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/method_comparison/MetaMathQA/experiments/lora/llama-3.2-3B-rank32/adapter_config.json))
- NoRA: 0.479 (same config as LoRA, except setting `use_nora=True`)
- BIMI: 0.471 (same config as LoRA, except setting `init_lora_weights="bimi"`)

I also checked the train loss curves, there is an interesting pattern where NoRA and BIMI start off better, then normal LoRA catches up, then falls back again:

Image

It would be great if you could share more details.

As for the implementations:

1. Could you please create a separate PR for BIMI and NoRA, as they're orthogonal?
2. BIMI init is not documented or tested, please add those.
3. For NoRA, we need to implement this as a LoRA variant. In the `forward` and `get_delta_weight` method, we don't want to have any `if self.use_nora.get(active_adapter) is True`. The variant should be pretty simple.
4. It looks like you want to enable NoRA also for DoRA. This should be implemented as a "DoraNora" variant. Then, in the [`lora_variants` property](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/src/peft/tuners/lora/layer.py#L936-L950), you can add a tuple `("usa_dora", "use_nora"): variants.DoraNoraLinearVariant` etc.
5. For NoRA, let's add it to the test matrix, same as DoRA, to ensure it works. If you add the "DoraNora" variant, include it in the test matrix too. Check the [entries for DoRA](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/tests/test_custom_models.py#L151) for reference.

**Comment by Joluck** [comment]
> Thanks for integrating these new options into LoRA.
> 
> To see the results, I ran the MetaMathQA/GSM8K benchmark in PEFT. I'm not exactly sure if the experiment is directly comparable to the one in the paper, do you have the code for that somewhere? The end result for me was that the test accuracy for LoRA, NoRA, and BIMI are very close, within the noise range (all rank 32):
> 
> * LoRA: [0.478](https://github.com/huggingface/peft/blob/05456023a231632c1de47b23edafa7e5adf86148/method_comparison/MetaMathQA/results/lora--llama-3.2-3B-rank32.json#L334) (with [this config](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/method_comparison/MetaMathQA/experiments/lora/llama-3.2-3B-rank32/adapter_config.json))
> * NoRA: 0.479 (same config as LoRA, except setting `use_nora=True`)
> * BIMI: 0.471 (same config as LoRA, except setting `init_lora_weights="bimi"`)
> 
> I also checked the train loss curves, there is an interesting pattern where NoRA and BIMI start off better, then normal LoRA catches up, then falls back again:
> 
> Image
> It would be great if you could share more details.
> 
> As for the implementations:
> 
> 1. Could you please create a separate PR for BIMI and NoRA, as they're orthogonal?
> 2. BIMI init is not documented or tested, please add those.
> 3. For NoRA, we need to implement this as a LoRA variant. In the `forward` and `get_delta_weight` method, we don't want to have any `if self.use_nora.get(active_adapter) is True`. The variant should be pretty simple.
> 4. It looks like you want to enable NoRA also for DoRA. This should be implemented as a "DoraNora" variant. Then, in the [`lora_variants` property](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/src/peft/tuners/lora/layer.py#L936-L950), you can add a tuple `("usa_dora", "use_nora"): variants.DoraNoraLinearVariant` etc.
> 5. For NoRA, let's add it to the test matrix, same as DoRA, to ensure it works. If you add the "DoraNora" variant, include it in the test matrix too. Check the [entries for DoRA](https://github.com/huggingface/peft/blob/e1fddd9bb428f0f2207c73f095a562b5bc6d2b3e/tests/test_custom_models.py#L151) for reference.

The SFT experiment can be directly verified at https://github.com/huggingface/peft/pull/3637#pullrequestreview-5113952209. Note that NoRA has two modes: continuous norm constraint and single init constraint. BIMI is a special case of NoRA-init. One more point to note: I recommend setting rank=alpha, so that the initial gradients of NoRA and full fine-tuning are basically the same (of course, rank:alpha=1:2 can achieve a lower loss in certain domains or with less training, but it also brings more forgetting).

**Comment by Joluck** [comment]
> > 3. For NoRA, we need to implement this as a LoRA variant. In the `forward` and `get_delta_weight` method, we don't want to have any `if self.use_nora.get(active_adapter) is True`. The variant should be pretty simple.

NoRA can impose a continuous constraint on A. Personally, I also prefer that users adopt the simpler NoRA-init (which is straightforward, effective, adds no extra overhead, and is suitable for frameworks like vLLM). Whether to include the standard NoRA in PEFT depends on your decision to adopt it.

**Comment by BenjaminBossan** [comment]
> The SFT experiment can be directly verified at [#3637 (review)](https://github.com/huggingface/peft/pull/3637#pullrequestreview-5113952209)

What do you mean by that? If you have experiment configs for NoRA and BIMI, please include them in the corresponding PRs.

> Note that NoRA has two modes: continuous norm constraint and single init constraint.

I only tried `True` so far, as this should have the strongest effect.

> Personally, I also prefer that users adopt the simpler NoRA-init (which is straightforward, effective, adds no extra overhead, and is suitable for frameworks like vLLM). Whether to include the standard NoRA in PEFT depends on your decision to adopt it.

We could also start with a PR that offers only the NoRA init option as a simple `init_lora_weights="nora"` option, same as with BIMI. If this already covers most of the effect and, as you mention, makes downstream adoption easier, it makes most sense.

**Comment by BenjaminBossan** [comment]
Gentle ping @Joluck

**Comment by Joluck** [comment]
> > The SFT experiment can be directly verified at [#3637 (review)](https://github.com/huggingface/peft/pull/3637#pullrequestreview-5113952209)
> 
> What do you mean by that? If you have experiment configs for NoRA and BIMI, please include them in the corresponding PRs.

The compare for PEFT has too little training, so no significant difference can be seen.


> 
> > Note that NoRA has two modes: continuous norm constraint and single init constraint.
> 
> I only tried `True` so far, as this should have the strongest effect.
> 
> > Personally, I also prefer that users adopt the simpler NoRA-init (which is straightforward, effective, adds no extra overhead, and is suitable for frameworks like vLLM). Whether to include the standard NoRA in PEFT depends on your decision to adopt it.
> 
> We could also start with a PR that offers only the NoRA init option as a simple `init_lora_weights="nora"` option, same as with BIMI. If this already covers most of the effect and, as you mention, makes downstream adoption easier, it makes most sense.


Sorry, I've been too busy lately. I'll submit an implementation of `init_lora_weights="nora"` as soon as possible.

**Comment by BenjaminBossan** [comment]
> Sorry, I've been too busy lately. I'll submit an implementation of `init_lora_weights="nora"` as soon as possible.

Thanks! No need to rush.
